DOC: v3 API docs (#718)
imageio/imageio: docs/reference/userapi.rst

.. autoclass:: imageio.core.format.Writer
    :inherited-members:
    :members: 

.. _migration_from_v2:

Migrating to ImageIO v3
^^^^^^^^^^^^^^^^^^^^^^^

The primary novelty of the v3 API compared to the v2 API is that images are now
treated as ndimages, a generalization of (flat) 2D images to N dimensions. This
change allows us to design a much simpler API; however, comes with a few
backwards incompatible changes.

Avoiding API Migration for Legacy Scripts
-----------------------------------------

If you have an old script where you do not wish to migrate the API, you can
explicitly import the old v2 API and avoid migration. Note that, while this will
keep the old (v2) API, it will use on the new (v3) plugins and backends, which
may contain other backwards incompatible changes. As such, you should prefer a
full migration to v3, and should avoid using the v2 API in new code. To import
the v2 API, depending on your script/code, use::

    import imageio... as iio

>     import imageio... as imageio

Reviewer: This is invalid syntax and looks confusing. Do you mean `import imageio.v2 as iio`?

Author: Yes, I agree. This is still a TODO.  https://github.com/imageio/imageio/pull/720 is not yet merged, and its exact layout determines how this should be imported here.

Will fix it once we are clear on our transition policy.

Author: I changed it to `import imageio.v2 as iio` for now. If we end up agreeing on something else in #720 I will update again.

order they appear. Further, the default behavior of `iio.imread` has changed,
and it now returns the first ndimage (instead of the first flat image) if the
image contains multiple images.

To reproduce similar behavior to the v2 API behavior use one of the following
snippets::

    # img = iio.v2.imread(image_resource)
    try:
        img = iio.v3.imread(image_resource, index=None)[0]
    except ValueError:
        img = iio.v3.imread(image_resource, index=0)

    # img = iio.v2.volread(image_resource)
    img = iio.v3.imread(image_resource)
    
    # img = iio.v2.mimread(image_resource)
    # img = iio.v2.mvolread(image_resource)
    img = [im for im in iio.v3.imiter(image_resource)]

Writing Images
--------------

Similar to reading images, the new `iio.imwrite` can handle ndimages.
``iio.mimwrite``, ``iio.volwrite``, and ``iio.mvolwrite`` have all dissapeared.
The same goes for their aliases ``iio.mimsave``, ``iio.volsave``,
``iio.mvolsave``, and ``iio.imsave``. They are now all covered by
``iio.imwrite``.

To reproduce similar behavior to the v2 API behavior use one of the following
snippets::

    # img = iio.v2.imwrite(image_resource, ndimage)

>     # img = iio.v2.imsave(image_resource, ndimage)

Reviewer: Maybe don't mention the aliases here to avoid verbosity.

Author: Will it be confusing for people that use the alias to not find it in the list? Or will us mentioning the aliases in the text above suffice?

Reviewer: I think they'll get the point because the pattern is quite clear, and we mention the aliases above.

(reading) or  ``w`` (writing) for each `ImageResource` being opened.
As such, the functions ``iio.get_reader``, ``iio.get_writer``, ``iio.read``, and
``iio.save`` have become obsolete and the plugin object is used directly
instead. Using these is considered advanced usage, happens inside a context
manager, and uses `iio.imopen`::

    # reader = iio.get_reader(image_resource)
    # reader = iio.read(image_resource)
    with iio.imopen(image_resource, "r") as reader:
        # file is only open/accessible here
        # img = reader.read(index=0)
        # meta = reader.metadata(index=0)
        # ...
        pass

    # writer = iio.get_writer(image_resource)
    # writer = iio.save(image_resource)
    with iio.imopen(image_resource, "w") as writer:
        # file is only open/accessible here
        # writer.write(ndimage)
        # ...
        pass

Metadata
--------

The v3 API makes handling of metadata much more consistent and explicit.
Previously in v2, metadata was provided as a ``dict`` that was attached to the
image by returning a custom subclass of ``np.ndarray``. Now metadata is
provided separately from pixel data::

>     # metadata = iio.imread(image_resource)._meta

Reviewer: ```suggestion
    # metadata = iio.imread(image_resource).meta
```

    metadata = iio.immeta(image_resource)

Further, ImageIO now provides a curated set of standardized metadata which is
called ImageProperties in addition to above metadata. The difference between the
two is as follows: The metadata dict contains metadata using format-specific
keys to the extent the reading plugin supports them. The ImageProperties

> dataclass contains metadata using standardized attribute names. They are the
> same name for all plugins, and - if the plugin or format doesn't provide a field
> - they are set to a (sensible) default value; usually ``None``. To access

Reviewer: ```suggestion
dataclass contains metadata using standardized attribute names. Each plugin
provides the same set of properties, and if the plugin or format doesn't provide a field
it is set to a (sensible) default value. To access
```

    try:
        img = iio.v3.imread(image_resource, index=None)[0]
    except ValueError:
        img = iio.v3.imread(image_resource, index=0)

    # img = iio.v2.volread(image_resource)
    img = iio.v3.imread(image_resource)
    
    # img = iio.v2.mimread(image_resource)
    # img = iio.v2.mvolread(image_resource)
    img = [im for im in iio.v3.imiter(image_resource)]

Writing Images
--------------

Similar to reading images, the new `iio.imwrite` can handle ndimages.
``iio.mimwrite``, ``iio.volwrite``, and ``iio.mvolwrite`` have all dissapeared.
The same goes for their aliases ``iio.mimsave``, ``iio.volsave``,
``iio.mvolsave``, and ``iio.imsave``. They are now all covered by
``iio.imwrite``.

To reproduce similar behavior to the v2 API behavior use one of the following
snippets::

    # img = iio.v2.imwrite(image_resource, ndimage)
    # img = iio.v2.volwrite(image_resource, ndimage)
    img = iio.v3.imwrite(image_resource, ndimage)

    # img = iio.v2.mimwrite(image_resource, list_of_ndimages)
    # img = iio.v2.mvolwrite(image_resource, list_of_ndimages)
    img = iio.v3.imwrite(image_resource, list_of_ndimages)

Reader and Writer
-----------------

> Plugins now longer have nested ``Reader`` and ``Writer`` classes and they are no
> longer global singletons. Instead, a new plugin object is instantiated in mode ``r``
> (reading) or  ``w`` (writing) for each `ImageResource` being opened.
> As such, the functions ``iio.get_reader``, ``iio.get_writer``, ``iio.read``, and
> ``iio.save`` have become obsolete and the plugin object is used directly
> instead. Using these is considered advanced usage, happens inside a context
> manager, and uses `iio.imopen`::

Reviewer: The mentioning of nested classes and singletons is not relevant for normal users, so let's omit that.

```suggestion
Previously, the `reader` and `writer` objects provided an advanced way to read/write data with more control. These are now removed. Likewise, the functions ``iio.get_reader``, ``iio.get_writer``, ``iio.read``, and
``iio.save`` have become obsolete.

Instead, the plugin object is used directly, which is instantiated in mode ``r`` (reading) or  ``w`` (writing).
To create a plugin object, call `iio.imopen` and use it as a context manager::
```

        # file is only open/accessible here
        # img = reader.read(index=0)
        # meta = reader.metadata(index=0)
        # ...
        pass

    # writer = iio.get_writer(image_resource)
    # writer = iio.save(image_resource)
    with iio.imopen(image_resource, "w") as writer:
        # file is only open/accessible here
        # writer.write(ndimage)
        # ...
        pass

Metadata
--------

The v3 API makes handling of metadata much more consistent and explicit.
Previously in v2, metadata was provided as a ``dict`` that was attached to the
image by returning a custom subclass of ``np.ndarray``. Now metadata is
served independently from pixel data::

    # metadata = iio.imread(image_resource).meta
    metadata = iio.immeta(image_resource)

Further, ImageIO now provides a curated set of standardized metadata, which is
called :class:`ImageProperties <imageio.core.v3_plugin_api.ImageProperties>`, in
addition to above metadata. The difference between the two is as follows: The
metadata dict contains metadata using format-specific keys to the extent the
reading plugin supports them. The ImageProperties dataclass contains generally

> available metadata using standardized attribute names. Each plugin provides the

Reviewer: Does this link work? Links in rst always confuse me.
```suggestion
available metadata using standardized attribute names (check the `ImageProperties docs <imageio.core.v3_plugin_api.ImageProperties>`_). Each plugin provides the
```

Author: Yes it does. Broken links trigger a warning in sphinx, and we have both a CI job that builds the docs (with the same settings as CD) as well as RTD building PR docs we can inspect. I've configured both to fail on warnings so that we can quickly spot things like broken links ... useful feature :D

same set of properties, and if the plugin or format doesn't provide a field it

> is set to a (sensible) default value. To access ImageProperties use::

Reviewer: or to None, still TBD

Author: True. My thinking here was that "sensible default value" admits to the interpretation of: "it is always None if not provided", so we we are correct in either case x)

Reviewer: Ok, but would still be good to mention that values can be `None` so the user can take that into account. If we go that route.

Author: agreed. Does this mean we have a decision for https://github.com/imageio/imageio/issues/754? 

Either way, I have changed this around and added the phrase "; usually `None`". I think this is as close as we get until we make the definite decision to return `None` when the value is not set.

Reviewer: Yes, I think it makes sense.

Author: Perfect. That was the last open point in this review. Any other points or shall we merge?
